File: http/src/main/java/com/xinzy/http/CookieStore.java
package com.xinzy.http;

import android.text.TextUtils;
import android.util.ArrayMap;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import okhttp3.Cookie;
import okhttp3.HttpUrl;

/**
 * Created by Xinzy on 2017/6/9.
 */

interface CookieStore {
    void save(HttpUrl url, List<Cookie> cookies);

    List<Cookie> load(HttpUrl url);


    class MemoryCookieStore implements CookieStore {
        private Map<String, List<Cookie>> mCookies;

        MemoryCookieStore() {
            mCookies = Collections.synchronizedMap(new ArrayMap<String, List<Cookie>>());
        }

        @Override
        public void save(HttpUrl url, List<Cookie> cookies) {
            Utils.debug("MemoryCookieStore: cookie = " + cookies);
            mCookies.put(url.host(), cookies);
        }

        @Override
        public List<Cookie> load(HttpUrl url) {
            String host = url.host();
            List<Cookie> cookies = mCookies.get(host);
            if (cookies != null && cookies.size() > 0) {
                List<Cookie> cs = new ArrayList<>();
                for (Cookie cookie : cookies) {
                    if (!Utils.isCookieExpired(cookie)) {
                        cs.add(cookie);
                    }
                }
                return cs;
            }

            return new ArrayList<>(0);
        }
    }

    class PersistentCookieStore implements CookieStore {
        private File mCookieDir;
        private Gson mGson;

        PersistentCookieStore(File dir) {
            this.mCookieDir = dir;
            if (!mCookieDir.exists()) {
                mCookieDir.mkdirs();
            }
            mGson = new Gson();
        }

        @Override
        public void save(HttpUrl url, List<Cookie> cookies) {
            if (cookies != null && cookies.size() > 0) {
                File file = new File(mCookieDir, Utils.md5(url.host()));

                List<C> cs = new ArrayList<>();
                for (Cookie cookie : cookies) {
                    cs.add(C.convert(cookie));
                }
                String content = mGson.toJson(cs);
                Utils.debug("PersistentCookieStore: cookie = " + content);
                Utils.write(file, content);
            }
        }

        @Override
        public List<Cookie> load(HttpUrl url) {
            File f = new File(mCookieDir, Utils.md5(url.host()));
            String content = Utils.read(f);
            List<Cookie> cookies = new ArrayList<>();
            if (!TextUtils.isEmpty(content)) {
                try {
                    List<C> list = mGson.fromJson(content, new TypeToken<List<C>>(){}.getType());
                    if (list != null && list.size() > 0) {
                        for (C c : list) {
                            Cookie cookie = c.convert();
                            if (!Utils.isCookieExpired(cookie)) {
                                cookies.add(cookie);
                            }
                        }
                    }
                } catch (Exception e) {
                }
            }
            Utils.debug("PersistentCookieStore.load cookie: " + cookies);
            return cookies;
        }
    }

    class C {
        public String name;
        public String value;
        public long expiresAt;
        public String domain;
        public String path;
        public boolean secure;
        public boolean httpOnly;

        public boolean persistent;
        public boolean hostOnly;

        C(String name, String value, long expiresAt, String domain, String path,
                 boolean secure, boolean httpOnly, boolean persistent, boolean hostOnly) {
            this.name = name;
            this.value = value;
            this.expiresAt = expiresAt;
            this.domain = domain;
            this.path = path;
            this.secure = secure;
            this.httpOnly = httpOnly;
            this.persistent = persistent;
            this.hostOnly = hostOnly;
        }

        static C convert(Cookie cookie) {
            return new C(cookie.name(), cookie.value(), cookie.expiresAt(), cookie.domain(), cookie.path(),
                    cookie.secure(), cookie.httpOnly(), cookie.persistent(), cookie.hostOnly());
        }

        Cookie convert() {
            Cookie.Builder builder = new Cookie.Builder().name(name).value(value).expiresAt(expiresAt).domain(domain).path(path);
            if (secure) {
                builder.secure();
            }
            if (hostOnly) {
                builder.httpOnly();
            }
            return builder.build();
        }
    }
}

File: httpkotlin/src/main/java/com/xinzy/http/kotlin/CookieStore.kt
package com.xinzy.http.kotlin

import android.text.TextUtils

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import java.io.File

import okhttp3.Cookie
import okhttp3.HttpUrl

/**
 * Created by Xinzy on 2017/6/9.
 */

internal interface CookieStore {
    fun save(url: HttpUrl, cookies: List<Cookie>)
    fun load(url: HttpUrl): List<Cookie>
}

internal class MemoryCookieStore internal constructor() : CookieStore {
    private val mCookies = mutableMapOf<String, List<Cookie>>()

    override fun save(url: HttpUrl, cookies: List<Cookie>) {
        mCookies[url.host()] = cookies
    }

    override fun load(url: HttpUrl): List<Cookie> {
        val host = url.host()
        val cookies = mCookies[host]
        if (cookies != null && cookies.isNotEmpty()) {
            val cs = mutableListOf<Cookie>()
            cookies.filterNot { isCookieExpired(it) }.forEach { cs.add(it) }
            return cs
        }

        return listOf()
    }
}

internal class PersistentCookieStore internal constructor(private val mCookieDir: File) : CookieStore {
    private val mGson = Gson()

    init {
        if (!mCookieDir.exists()) {
            mCookieDir.mkdirs()
        }
    }

    override fun save(url: HttpUrl, cookies: List<Cookie>) {
        if (cookies.isNotEmpty()) {
            val cs = mutableListOf<C>()
            cookies.forEach { cs.add(C.convert(it)) }
            File(mCookieDir, md5(url.host())).writeText(mGson.toJson(cs))
        }
    }

    override fun load(url: HttpUrl): List<Cookie> {
        val content = File(mCookieDir, md5(url.host())).readText()
        val cookies = mutableListOf<Cookie>()
        if (!TextUtils.isEmpty(content)) {
            try {
                val list = mGson.fromJson<List<C>>(content, object : TypeToken<List<C>>() {}.type)
                list?.let { cs -> run {
                    cs.filter { it.isExpired() }.forEach { cookies.add(it.convert()) }
                } }
            } catch (e: Exception) {
            }
        }
        return cookies
    }
}

internal class C internal constructor(private var name: String, private var value: String, private var expiresAt: Long,
                                      private var domain: String, private var path: String, private var secure: Boolean,
                                      private var httpOnly: Boolean, private var persistent: Boolean,
                                      private var hostOnly: Boolean) {

    internal fun convert(): Cookie {
        val builder = Cookie.Builder().name(name).value(value).expiresAt(expiresAt).domain(domain).path(path)
        if (secure) {
            builder.secure()
        }
        if (httpOnly) {
            builder.httpOnly()
        }
        return builder.build()
    }

    fun isExpired() = isCookieExpired(convert())

    companion object {
        internal fun convert(cookie: Cookie): C {
            return C(cookie.name(), cookie.value(), cookie.expiresAt(), cookie.domain(), cookie.path(),
                    cookie.secure(), cookie.httpOnly(), cookie.persistent(), cookie.hostOnly())
        }
    }
}